### ISSUE TITLE
perl: time-related test failues on clang32 (likely 32-bit UCRT related)

### ISSUE THREAD

**Issue by jeremyd2019** [msys2/MINGW-packages#9178]
The following are differences between mingw32 and clang32 perl tests:
```diff
545c547,554
< io/fs.t ............................................................ ok
---
> # Failed test 28 - utime at io/fs.t line 250
> #      got "0"
> # expected "1"
> # Failed test 30 - atime: default case at io/fs.t line 529
> # Failed test 31 - mtime: default case at io/fs.t line 530
> io/fs.t ............................................................ 
> Failed 3/61 subtests 
> 	(less 20 skipped subtests: 38 okay)
552c561,564
< io/open.t .......................................................... ok
---
> # Failed test 149 - set mtime to a known value at io/open.t line 439
> io/open.t .......................................................... 
> Failed 1/188 subtests 
> 	(less 29 skipped subtests: 158 okay)
1029a1042,1136
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 458.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 458.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 445.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 458.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 487.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 761.
> Could not update timestamp at t/02_methods.t line 500.
> Could not update timestamp at t/02_methods.t line 529.
> Could not update timestamp at t/02_methods.t line 529.
> Could not update timestamp at t/02_methods.t line 529.
> Could not update timestamp at t/02_methods.t line 529.
> Could not update timestamp at t/02_methods.t line 542.
1045a1153,1160
> Could not update timestamp at t/04_resolved_issues.t line 148.
> Could not update timestamp at t/04_resolved_issues.t line 221.
> Could not update timestamp at t/04_resolved_issues.t line 221.
> Could not update timestamp at t/04_resolved_issues.t line 221.
> Could not update timestamp at t/04_resolved_issues.t line 235.
> Could not update timestamp at t/04_resolved_issues.t line 235.
> Could not update timestamp at t/04_resolved_issues.t line 235.
> Could not update timestamp at t/04_resolved_issues.t line 274.
1237c1352,1356
< ../cpan/ExtUtils-Install/t/Installapi2.t ........................... ok
---
> utime 'blib/lib/Big/Dummy.pm' failed:Invalid argument at t/Installapi2.t line 202.
> # Looks like your test exited with 22 just after 55.
> ../cpan/ExtUtils-Install/t/Installapi2.t ........................... 
> Dubious, test returned 22 (wstat 5632, 0x1600)
> Failed 15/70 subtests 
1321d1439
< ../cpan/HTTP-Tiny/t/010_url.t ...................................... ok
1328a1447
> ../cpan/HTTP-Tiny/t/010_url.t ...................................... ok
1362c1481,1491
< ../cpan/ExtUtils-MakeMaker/t/eu_command.t .......................... ok
---
> 
> #   Failed test 'eqtime'
> #   at t/eu_command.t line 84.
> #     '2'
> #         <=
> #     '1'
> # Looks like you failed 1 test of 40.
> ../cpan/ExtUtils-MakeMaker/t/eu_command.t .......................... 
> Dubious, test returned 1 (wstat 256, 0x100)
> Failed 1/40 subtests 
> 	(less 8 skipped subtests: 31 okay)
1425a1555,1584
> 
> #   Failed test 'mirror-01.txt request'
> #   at t/110_mirror.t line 76.
> #          got: 'GET /modified.txt HTTP/1.1^M
> # Connection: close^M
> # Host: example.com^M
> # If-Modified-Since: Tue, 20 Jul 2021 00:47:58 GMT^M
> # User-Agent: HTTP-Tiny/0.076^M
> # '
> #     expected: 'GET /modified.txt HTTP/1.1^M
> # Connection: close^M
> # Host: example.com^M
> # If-Modified-Since: Tue, 01 Feb 1994 00:00:00 GMT^M
> # User-Agent: HTTP-Tiny/0.076^M
> # '
> 
> #   Failed test 'mirror-02.txt request'
> #   at t/110_mirror.t line 76.
> #          got: 'GET /not-modified.txt HTTP/1.1^M
> # Connection: close^M
> # Host: example.com^M
> # If-Modified-Since: Tue, 20 Jul 2021 00:47:58 GMT^M
> # User-Agent: HTTP-Tiny/0.076^M
> # '
> #     expected: 'GET /not-modified.txt HTTP/1.1^M
> # Connection: close^M
> # Host: example.com^M
> # If-Modified-Since: Tue, 01 Feb 1994 00:00:00 GMT^M
> # User-Agent: HTTP-Tiny/0.076^M
> # '
1428c1587,1590
< ../cpan/HTTP-Tiny/t/110_mirror.t ................................... ok
---
> # Looks like you failed 2 tests of 20.
> ../cpan/HTTP-Tiny/t/110_mirror.t ................................... 
> Dubious, test returned 2 (wstat 512, 0x200)
> Failed 2/20 subtests 
1455a1618,1619
> Can't utime('1626741191', '1608285506', 'T:\mingw-w64-perl\src\build-i686-perl-5.32.1\cpan\autodie\t\touch_me'): Invalid argument at t/utime.t line 24
> # Looks like your test exited with 22 just after 4.
1463c1627,1629
< ../cpan/autodie/t/utime.t .......................................... ok
---
> ../cpan/autodie/t/utime.t .......................................... 
> Dubious, test returned 22 (wstat 5632, 0x1600)
> All 4 subtests passed 
3343c3510,3519
< ../lib/Time/localtime.t ............................................ ok
---
> 
> #   Failed test 'localtime(2147483647)'
> #   at ../lib/Time/localtime.t line 19.
> #          got: '0'
> #     expected: '9'
> Can't call method "sec" on an undefined value at ../lib/Time/localtime.t line 21.
> # Looks like your test exited with 22 just after 72.
> ../lib/Time/localtime.t ............................................ 
> Dubious, test returned 22 (wstat 5632, 0x1600)
> Failed 50/121 subtests 
3356a3533,3539
> io/fs.t                                                          (Wstat: 0 Tests: 61 Failed: 3)
>   Failed tests:  28, 30-31
> io/open.t                                                        (Wstat: 0 Tests: 188 Failed: 1)
>   Failed test:  149
> ../cpan/ExtUtils-Install/t/Installapi2.t                         (Wstat: 5632 Tests: 55 Failed: 0)
>   Non-zero exit status: 22
>   Parse errors: Bad plan.  You planned 70 tests but ran 55.
3363a3547,3554
> ../cpan/ExtUtils-MakeMaker/t/eu_command.t                        (Wstat: 256 Tests: 40 Failed: 1)
>   Failed test:  9
>   Non-zero exit status: 1
> ../cpan/HTTP-Tiny/t/110_mirror.t                                 (Wstat: 512 Tests: 20 Failed: 2)
>   Failed tests:  1, 5
>   Non-zero exit status: 2
> ../cpan/autodie/t/utime.t                                        (Wstat: 5632 Tests: 4 Failed: 0)
>   Non-zero exit status: 22
3372c3563,3567
< Files=2678, Tests=1153655, 1202 wallclock secs (98.02 usr + 13.48 sys = 111.50 CPU)
---
> ../lib/Time/localtime.t                                          (Wstat: 5632 Tests: 72 Failed: 1)
>   Failed test:  72
>   Non-zero exit status: 22
>   Parse errors: Bad plan.  You planned 121 tests but ran 72.
> Files=2678, Tests=1153731, 1205 wallclock secs (97.73 usr + 13.36 sys = 111.09 CPU)
```

These occur whether or not `_USE_32BIT_TIME_T` is defined.

**Comment by jeremyd2019** [comment]
It appears that it is in fact using 32-bit `time_t`.  I did a bit of debugging on the last error, and apparently `time_t t =2147483647; 
 printf("%p\n", _localtime32(&t));` returns NULL.  That doesn't answer why any of the other tests failed though...

**Comment by jeremyd2019** [comment]
First test is due to `utime` failing.  Apparently it's calling `_utime64` under the hood, but with 32-bit time_ts?  Don't know how that happens, but...  Seems like a mingw-w64 bug, so have reported to their mailing list.

**Comment by jeremyd2019** [comment]
> apparently `time_t t =2147483647; printf("%p\n", _localtime32(&t));` returns NULL. 

from the ucrt source in windows sdk, ucrt/inc/corecrt_internal_time.h has, for 32-bit time_t:
max_time_t = 0x7fffd27f

So this perl test is failing because it is working with 32-bit `time_t` and hitting that limit.

The other perl tests seem to be failing because it's calling `_utime64` with 32-bit `time_t` and returning -1/`errno`=`EINVAL`

**Comment by jeremyd2019** [comment]
[Proposed patch on mingw-w64-public mailing list](https://sourceforge.net/p/mingw-w64/mailman/message/37325459/) causes clang32 and clang64 tests outcomes to match (and before merging #9164 I made sure that clang64 and ucrt64 results matched mingw64).

**Comment by jeremyd2019** [comment]
https://sourceforge.net/p/mingw-w64/mingw-w64/ci/0e0eed62328177b31a54023d9e3dad7bc12c0fd0/

Eep.  @lazka do you think we can just update mingw-w64 packages, or will this require a lot of clang32 rebuilding for ABI breakage?

**Comment by jeremyd2019** [comment]
This was fixed in #9259
